### PR TITLE
DOC: SciPy extensions for code style and docstring guidelines.

### PR DESCRIPTION
**Pull request by WarrenWeckesser**
I've been keeping notes on various requests for changes that come up now and then in PRs in NumPy and SciPy.   In this PR, I've documented some of the common requests that I've seen in a new section of the docs called "Code and Documentation Style Guide - The Missing Bits".   These are issues that are not covered in the usual coding and documentation PEPs and guides, and that can be expressed as a fairly simple rule.

The motivation for these guidelines is this: if a core developer makes a request to change a PR that is a *blocking* request, and if the change can be expressed as a fairly simple rule, and if there is consensus among the core devs that the requested change *should* block the PR, then the rule should be documented.  That way it is clear that the requested change is not just the reviewer's personal preference, but in fact is the preferred style approved by the SciPy devs.

It is unlikely that everyone will agree that we need all these new guidelines.  In fact, I expect that for some of these guidelines, a reaction will be "it is not important, we don't need it".  That's fine!  But then we have to be sure that in the future, if a reviewer requests a such a change in a PR, the request is just a suggestion (i.e. an expression of personal preference), and *not* a request that will block the PR.

For example, I personally don't care about the relatively trivial guidelines about where to put the space when a long string is broken at a word boundary, or about the suggestion to always include a blank line as the last line of a multiline docstring.  If everyone else feels the same, then I'll take those out.  But that means reviewers who *do* happen to have a preference should not block a PR that doesn't follow the guidelines. (There is nothing wrong with stating the preference in a review comment, but it shouldn't block a PR.)

### CODE REVIEW

File: doc/source/dev/missing-bits.rst
Comment (tylerjereddy) [inline doc/source/dev/missing-bits.rst:43]: Seems useful--perhaps not the "prettiest" function signature. My brain tends to immediately think of unpacking or dereferencing when I see that.

File: doc/source/dev/missing-bits.rst
Comment (tylerjereddy) [inline doc/source/dev/missing-bits.rst:111]: oh dear!

File: doc/source/dev/missing-bits.rst
Comment (tylerjereddy) [inline doc/source/dev/missing-bits.rst:132]: I don't think I've ever even considered blocking a PR for this differentiation. Using a `\` to continue will usually get me to say something though.

File: doc/source/dev/missing-bits.rst
Comment (tylerjereddy) [inline doc/source/dev/missing-bits.rst:260]: I've never really thought of blocking a PR for this either. Don't really mind it if you want to though.

File: doc/source/dev/missing-bits.rst
Comment (rgommers) [reply doc/source/dev/missing-bits.rst:43]: +1 for using keyword-only arguments more.

File: doc/source/dev/missing-bits.rst
Comment (rgommers) [reply doc/source/dev/missing-bits.rst:260]: I do mind, this is a typical example of being too nitpicky, which is quite unpleasant for new contributors. Just push a fix before merging for something this small.

**Review by rgommers — COMMENTED** [review]
Thanks for writing this up Warren. Most/all of this looks like good guidance, and useful to document. My one caution is that this should be documentation primarily for us (maintainers), you cannot expect new contributors to read and adhere to all that. And if they don't, keep in mind https://numpy.org/devdocs/dev/reviewer_guidelines.html#communication-guidelines. These are important to not make contributing to SciPy feel painful.

Related: [pre-commit](https://pre-commit.com/) is now much more user-friendly than manual pre-commit hooks were in the past, so integrating that and making it run `tools/lint_diff.py` (same as in CI) would also be useful I think. I haven't used it much, but quite a few people have told me they like it.

File: doc/source/dev/missing-bits.rst
Comment (rgommers) [reply doc/source/dev/missing-bits.rst:43]: This actually feels like it belongs somewhere more prominent. We typically forget, and it's an important API design rule.

File: doc/source/dev/missing-bits.rst
Comment (ev-br) [inline doc/source/dev/missing-bits.rst:21]: My common gripe is space after comma in lists and such. We do not enforce it in CI and IMO we should not --- this is a recommendation not a strict rule --- hence these extra guidelines feel like a perfect place for this. Suggested phrasing may be along the lines:

```
When writing lists and enumerations, please make sure to use spaces after commas.

Yes: `a = np.array([1, 2, 3], dtype=float)`
No: `a = np.array([1,2,3],dtype=float)`
```

File: doc/source/dev/missing-bits.rst
Comment (ev-br) [inline doc/source/dev/missing-bits.rst:142]: How about something along the lines of


When writing out formulas, consider using pseudocode or TeX markup. The former is preferred for simple short formulas but gets unwieldy for more complicated ones. In the latter case, you may use TeX markup instead. When using TeX markup, please consider TeX markup as souce code --- hence aim for readability. 
These are not hard and fast rules, more of general guidelines: 

- keep the number of curly braces to a minimum.
- avoid excessive markup control, e.g. `\left( `,  `\bigl` and so on.
- prefer simple, more readable constructs. For instance, consider using `/` instead of `\frac{...}{...}`

For instance

```
No:   $\frac{1}{\sqrt{2\pi\, \sigma^{2}}} e^{-\bigl(\frac{x - \mu}{2\sigma}\bigr)^2}$

Yes: $\exp(-(x-\mu)^2 / \sigma^2) / \sqrt(2\pi\sigma^2) $
```

File: doc/source/dev/missing-bits.rst
Comment (ev-br) [inline doc/source/dev/missing-bits.rst:164]: ESL here: does this wording mean that the function does handle negative values of `x`, and using `x<0` is merely not recommended?

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:132]: Agreed, I'm happy to remove this completely.  I included it because it has come up in a couple NumPy PRs.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:260]: @rgommers, that's fine.  I'm happy to remove this one, too.  Personally, I have no preference, and I don't think I would even bother pushing a "fix".  (I've probably written docstrings with and without a final blank line in the past.)

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:21]: This is already part of PEP-8, so it isn't really a "missing bit".

> We do not enforce it in CI and IMO we should not...

As a well-documented item in PEP-8, I think we *should* enforce it in CI, but with so much legacy code that does not conform to the standard, it is difficult to do.

**Comment by tupui** [comment]
> Related: [pre-commit](https://pre-commit.com/) is now much more user-friendly than manual pre-commit hooks were in the past, so integrating that and making it run `tools/lint_diff.py` (same as in CI) would also be useful I think. I haven't used it much, but quite a few people have told me they like it.

Also these tools now offer a way to not kill git blame and such. So having something like Black would remove lots of discussions. I set these up (black, pre-commit, isort) on a few projects at work and it's really convenient.

File: doc/source/dev/missing-bits.rst
Comment (tupui) [reply doc/source/dev/missing-bits.rst:21]: We could use black once and flag the commit for git blame (as per black's [doc](https://github.com/psf/black#migrating-your-code-style-without-ruining-git-blame)). Then the whole repo would be cleaned up.

File: doc/source/dev/missing-bits.rst
Comment (tupui) [inline doc/source/dev/missing-bits.rst:46]: We actually have a link in the doc pointing to NumPy's testing doc. And in NumPy's doc it's says their guide applies to us. We should put the link to NumPy here (https://docs.scipy.org/doc/numpy/reference/testing.html). And in `HACKING.rst.txt` and `development_workflow.rst`, maybe point to here as well.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:164]: I don't know what it means.  It *might* mean the alternative is not recommended but still accepted, or it might mean the code will raise an exception.   That's one reason we should prefer "must" over "should".  In most cases where I've seen "should" used, the intent appears to be the same as "must", and an error is raised if the condition is not satisfied.

In cases where the condition is, in fact, a recommendation but not something that is checked when the parameters are validated, then "should" (or something like it) is probably OK, but additional explanation is needed.  Perhaps something like this (where I've used "is expected to be" instead of "should be"):

```
        x : float
            x is expected be nonnegative.  Negative values are accepted, but the results
            are not meaningful in that case.
```

File: doc/source/dev/missing-bits.rst
Comment (tupui) [inline doc/source/dev/missing-bits.rst:180]: I would also add the preferred way to cite articles. Lately I have used: `Fang et al. "Design and modeling for computer experiments". Computer Science and Data Analysis Series, 2006.`

File: doc/source/dev/missing-bits.rst
Comment (tupui) [inline doc/source/dev/missing-bits.rst:204]: Speaking of examples, I would add the new way of not using seeds. I can write a suggestion here if you want?

**Review by tupui — COMMENTED** [review]
Nice, thanks for proposing that! (Do you have the feedback on skipping the CI?)

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:204]: Yes, that would be great!

File: doc/source/dev/missing-bits.rst
Comment (ev-br) [reply doc/source/dev/missing-bits.rst:21]: Re black and/or CI : IIRC that switch in pyflakes et al is way too strict and makes some algorithmic code less readable. There certainly were multiple discussions about that.

File: doc/source/dev/missing-bits.rst
Comment (tupui) [reply doc/source/dev/missing-bits.rst:132]: Yes instead I would talk about `\` and suggest that we prefer one of the examples you showed.

File: doc/source/dev/missing-bits.rst
Comment (tupui) [reply doc/source/dev/missing-bits.rst:21]: If this is a concern, there is always the option to use things like `# fmt: on/off` for instance.

Comment (ev-br) [reply doc/source/dev/missing-bits.rst:21]: These comments, with a sole reason being fending off over-eager linters look like line noise to a human. 
My two cents anyway

Comment (tupui) [reply doc/source/dev/missing-bits.rst:21]: Agreed, and maybe there are better ways like a general config. But the price being almost zero comments about style in review makes me want to dive in 😅

File: doc/source/dev/missing-bits.rst
Comment (tupui) [inline doc/source/dev/missing-bits.rst:205]:
```suggestion
How to use Random Number Generator (RNG) and seed
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In case random numbers are needed, `np.random.Generator` API must be used.
Use the helper function to validate the parameter `seed`, `random_state` [TBD: which is preferred?]::

    from scipy._lib._util import check_random_state

    def foo(seed=None):
        rng = check_random_state(seed)
        sample = rng.uniform(...)  # for instance

If you want to write an example using random numbers in the documentation::

    >>> rng = np.random.default_rng()
    >>> sample = rng.uniform(...)  # for instance

.. warning:: Do not specify the seed. This line will get overwritten and the same seed
   will be used for conveniency for all examples.

In tests, you have to use the `np.random.RandomState` interface and here you can use a seed
if you need a deterministic output::

    def test_foo():
        rng = np.random.RandomState(SOME_NUMBER)
        sample = rng.uniform(...)  # for instance

For further info, see `NumPy's documentation
<https://numpy.org/doc/stable/reference/random/index.html>`__.


```

File: doc/source/dev/missing-bits.rst
Comment (tupui) [reply doc/source/dev/missing-bits.rst:204]: Done bellow, hence resolving this one.

**Comment by ev-br** [comment]
One other thing to add here might be not using assets outside of test code. Errors need to be raised explicitly.

**Comment by tupui** [comment]
I am also seeing this in some PR as we add seeding, we should also decide about how to name this: `seed`, `random_sate`, something else?

File: doc/source/dev/missing-bits.rst
Comment (tupui) [inline doc/source/dev/missing-bits.rst:157]: Might be work talking about how to format the params in the description vs others things with double backticks.
```suggestion
                `x` must be nonnegative.
```

**Review by tupui — COMMENTED** [review]
I think we should prioritize this PR. Possibly backport it, but not necessary as we push people to look at the devdoc anyway. All these are quite important things that we always have to talk about during reviews (and we could add a link to this in the PR template).

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:180]: There are so many different conventions for this, and I have no preference, so for now I'm not going to try to provide general guidelines for the reference format.  Perhaps that can wait for a follow-up PR.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:193]: @tupui, do you think you could split this into two parts, one for generating random data in the "Examples" section of a docstring, and one for using random data in unit tests?  I hadn't followed very closely the discussions in https://github.com/scipy/scipy/pull/13863, but in https://github.com/scipy/scipy/pull/14182, @rkern pointed me to the important conclusions, including the part about a predefined seed being used when Sphinx builds the docs. This means a developer can test the code that will be used in an example, and be sure that the seed will match the one used in Sphinx.  We need to get some version of [this comment](https://github.com/scipy/scipy/pull/14182#issuecomment-855318533) into these docs.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst]: I have removed this guideline;  I don't think any of the SciPy devs care where the space goes, so we don't need it.

If it seems useful, we could add a note about "How to format a long string literal", that show our preference for using parentheses and implicit continuation instead of the backslash.  That could also be done in a follow-up PR.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst]: I removed the guideline about including a blank line at the end of a docstring.  Everyone should understand that the implication of *not* having such a guideline is that we generally leave it up to the PR author to do as they please.

**Comment by WarrenWeckesser** [comment]
Hmmm... why are the Azure tests running?  I thought `[skip azp]` was working these days?

**Comment by WarrenWeckesser** [comment]
> Hmmm... why are the Azure tests running? I thought [skip azp] was working these days?

Proposed fix for the handling of `[skip azp]` is in https://github.com/scipy/scipy/pull/14197.

File: doc/source/dev/missing-bits.rst
Comment (tupui) [reply doc/source/dev/missing-bits.rst:193]: Back end of June. I will get to this as soon as I'm back if this is still a thing.

**Comment by tupui** [comment]
This could interest you @melissawm.

File: doc/source/dev/missing-bits.rst
Comment (melissawm) [reply doc/source/dev/missing-bits.rst:49]: Please note that the new link is https://numpy.org/doc/stable/reference/testing.html

File: doc/source/dev/missing-bits.rst
Comment (melissawm) [inline doc/source/dev/missing-bits.rst:13]: I think this should point to the new url:
```suggestion
  <https://numpy.org/doc/stable/reference/testing.html>`_
```

File: doc/source/dev/missing-bits.rst
Comment (melissawm) [inline doc/source/dev/missing-bits.rst:61]: Same as before:
```suggestion
<https://numpy.org/doc/stable/reference/testing.html>`_
```

File: doc/source/dev/missing-bits.rst
Comment (ilayn) [reply doc/source/dev/missing-bits.rst:168]: Use of DOI must be emphasized over never-ending natbib or ieeetran discussions. DOIs are here to stay the rest is constantly changing.

Comment (tupui) [reply doc/source/dev/missing-bits.rst:168]: Agreed and I discovered recently we have special syntax for this and arXiv.

**Review by tupui — COMMENTED** [review]
@WarrenWeckesser are you still interested in continuing this?

**Comment by WarrenWeckesser** [comment]
> @WarrenWeckesser are you still interested in continuing this?

Yes--I'll revisit the doc this weekend.

**Comment by tupui** [comment]
> > @WarrenWeckesser are you still interested in continuing this?
> 
> Yes--I'll revisit the doc this weekend.

Thank you! It might also be worth updating our PR template and guidelines to indicate this page.

File: doc/source/dev/missing-bits.rst
Comment (mdhaber) [reply doc/source/dev/missing-bits.rst:134]: I think we need even more detail here. The [numpydoc style guide](https://numpydoc.readthedocs.io/en/latest/format.html) offers only:
image
image

We should at least mention that function names should also have single backticks because they get linked. (And that function names from other sub-packages should be fully-qualified, e.g. `scipy.optimize.minimize`.)

Examples of the ambiguity that remains:

- Sometimes return values of functions are named in the documentation. Are these "variables" or "inline display"?
- What about strings or other values that are passed as the value of an argument (e.g. `'exact'`, `None`, `1.5`)?
- Parameters that are accepted in `kwargs` or similar? Named as keys of an `options` structure?

Then there's the question about _why_ it's appropriate for parameter names, which _are_ ultimately code, to appear sometimes in bold:
image
sometimes in italics:
image
and sometimes as code (monospaced font / teletype):
image
(All of these are from [scipy.optimize.minimize](https://docs.scipy.org/doc/scipy/reference/generated/scipy.optimize.minimize.html#scipy.optimize.minimize), and all use single/double-backticks correctly, AFAICT.)

I'd suggest we can sidestep these questions and clarify the docs if all of these cases just appeared in monospace font, since they really are all code.

File: doc/source/dev/missing-bits.rst
Comment (tupui) [reply doc/source/dev/missing-bits.rst:82]: I would not recommend any use or `re` nor `capsys`. You should be able to do this use raw strings `match=r"...."`

Comment (mdhaber) [reply doc/source/dev/missing-bits.rst:82]: Use of `re` is a separate question, though.

I'm asking whether the f-string substitution should be matched to prevent issues like that one.

File: doc/source/dev/missing-bits.rst
Comment (mdhaber) [reply doc/source/dev/missing-bits.rst:134]: And @tupui noted [above](https://github.com/scipy/scipy/pull/13955#issuecomment-836645230), we should specify how to refer to code in Errors and Warnings, too. (Of course, I'd suggest single backticks here, too.)

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:130]: Even though I won't recommend that SciPy use it, there is a technical solution to the "LaTeX issue".  For my [`mpsci`](https://github.com/WarrenWeckesser/mpsci) repository, I created a Sphinx extension that uses regular expressions to replace text in a docstring when Sphinx is run.  For example, the text of the docstring for ``mpsci.fun.marcumq`` is

https://github.com/WarrenWeckesser/mpsci/blob/f524ed8ab3a4c84485e253b42c5c82c1030ab248/mpsci/fun/_marcumq.py#L14-L43

which is

```
def marcumq(m, a, b):
    """
    The Marcum Q function.

    The function is defined as the integral from b to infinity of

        x*(x/a)**(M-1)*exp(-(x**2 + a**2)/2)*I_{M-1}(a*x)

    where I_{M-1}(x) is the modified Bessel function of order M-1.

    *See also:* `mpsci.fun.cmarcumq`

    Returns
    -------
    q : mpmath.mpf
        The function value.

    Notes
    -----
    The function uses numerical integration, so it can be very slow.

    Examples
    --------
    >>> import mpmath
    >>> from mpsci.fun import marcumq
    >>> mpmath.mp.dps = 40
    >>> marcumq(2, 0.5, 3.0)
    mpf('0.07558718754263240906718640000640082605610073')

    """
```
There is no explicit LaTeX markup. (I guess  `I_{M-1}(x)` is LaTeXish, but it is not wrapped with the Sphinx `:math:` markup.) The HTML output, however, is [here](https://warrenweckesser.github.io/mpsci/fun/marcumq.html#mpsci.fun.marcumq), with integral signs, fractions, big parentheses, etc.

The reason I don't recommend it for SciPy is that maintaining the regular expressions is a fussy and fragile process.  It is easy to make a few changes to a docstring and forget to also update the regular expression that is used to match the text that is to be replaced with LaTeX markup.  I don't mind doing it in my one-man project, but I think it would be too much extra maintenance for SciPy.  I point it out just so folks can see what is *possible* (and maybe see if anyone can come up with a more robust method to do it).

Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:130]: I'm removing the section about the use of LaTeX from this PR.  If there is interest in continuing the discussion, it can be done in a follow-up PR.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:24]: I'm "resolving" this thread; to `black` or not to `black` is not something we should try to resolve in this PR.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:49]: The link to the numpy testing guidelines is a few lines down, and thanks to @melissawm, it now has the correct URL.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:46]: I'll leave this recommendation here. If it should be moved or copied somewhere else, I think we can do that in a follow-up PR.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:94]: Nice.  I'm not sure it would be worth the churn, however.  In fact, I'm removing this section from the PR.  If it turns out that there is interest in having an explicit guideline for the name to use for a parameter that is naturally called `lambda` (or perhaps more general guidelines for choosing parameter names), we can handle it in a follow-up PR.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:116]: Yeah, it looks bad when seen all at once like this, but a user reading the docstring of one of these functions would probably never notice the inconsistency.  I'm removing this section, but we can revisit the topic if there is interest.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:152]: Other than @ev-br's question, there have been no comments about this guideline, so I'll leave it as is and resolve this conversation.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:82]: That's an [interesting case](https://github.com/scipy/scipy/pull/15566/files/21f7971ddf0b611a3c802dc837cadbe7531681b0#diff-49422b768d9abc96ae02952d0fe0a68832b6ce211857185449ebb104cbe02681); the final fix that got merged was very simple, but there was no regression test added that would ensure that the fix worked.   That particular case is simple enough that it doesn't really bother me, and I don't know if similar situations are common enough to justify creating a guideline.  So all I can offer here is a shrug, and suggest we not attempt to impose written guidelines until we have more cases.

I also don't feel like I have anything to say about using pytest `capsys` at the moment.  I expect this "missing bits" document to evolve, so if patterns emerge that warrant additional guidelines for developers, we can add them later.

I'm leaving this conversation open, in case anyone else wants to chime in.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:168]: I'll add a note about recommending the use of DOIs, and mention the available markup.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst:117]: It has been roughly 11 months and no one has indicated a different preference, so let's go with the first option.

File: doc/source/dev/missing-bits.rst
Comment (tupui) [reply doc/source/dev/missing-bits.rst:156]: This has been updated and ready to include on my side.

**Comment by WarrenWeckesser** [comment]
I have updated the PR and moved it out of "draft" status.

These are unresolved items where conversation have been started here but would benefit from being split out as separate follow-up issues or PRs:
* Use of backticks around variable names: reconciling the NumPy docstring guidelines with (1) desired
  appearance of the rendered web pages, and (2) occasional incorrect links when single backticks are used.
* Use of delimiters around parameter names in error and warning messages.
* Using (and seeding) random numbers in docstrings and in unit tests. (@tupui has provided a starting point for these guidelines in the comments.)

I don't know what is happening with CircleCI; that's the one test that I wanted to run!

**Comment by ilayn** [comment]
We are using git+ssh instead of https and things changed on git side.

**Comment by tupui** [comment]
> I don't know what is happening with CircleCI; that's the one test that I wanted to run!

@WarrenWeckesser You need to merge main to this PR. Since you branched there has been a few changes which require that.

**Comment by tupui** [comment]
> These are unresolved items where conversations have been started here but would benefit from being split out as separate follow-up issues or PRs:
> 
> * Use of backticks around variable names: reconciling the NumPy docstring guidelines with (1) desired
>   appearance of the rendered web pages, and (2) occasional incorrect links when single backticks are used.

We already are enforcing something. We should describe the current state at least. It does not prevent us for discussing future changes.

> * Using (and seeding) random numbers in docstrings and in unit tests. (@tupui has provided a starting point for these 

I would not delay this. A lot of PR are wrong with respect to this and I always struggle to point out official doc for it.

**Review by rgommers — APPROVED** [review]
Everything that's here LGTM. +1 for merging as is, and leaving other topics for follow-ups.

**Review by mdhaber — CHANGES_REQUESTED** [review]
The rest LGTM, if we're holding other discussions for followups.

File: doc/source/dev/missing-bits.rst
Comment (WarrenWeckesser) [reply doc/source/dev/missing-bits.rst]: Done.

**Review by mdhaber — APPROVED** [review]

**Comment by WarrenWeckesser** [comment]
I forgot to add the `skip` notation in the last commit. The failed test is the Azure timeout that has been happening lately, and is not related to this PR.

File: doc/source/dev/missing-bits.rst
Comment (tupui) [reply doc/source/dev/missing-bits.rst:148]: I opened #15852 for that.

**Comment by ilayn** [comment]
This is already a very nice set of guidelines and we can fix the rough edges later. Already approved by maintainers and +1 from me too. In it goes thanks @WarrenWeckesser and all reviewers

**Comment by ilayn** [comment]
@melissawm @tylerjereddy there is already a review request for you but I think working on separate PRs would be better to avoid the review bloat on this, hence the merge.

**Pull request merged**
